Setup.hs

```diff
@@ -96,38 +97,65 @@ generateInterfaces pd lbi = do
   -- current directory root of the package.
 
   putStrLn "Generating Agda library interface files..."
-  forM_ (dataFiles pd) $ \fp -> when (takeExtension fp == ".agda") $ do
-    let fullpath  = ddir </> fp
-    let fullpathi = toIFile fullpath
 
-    -- remove existing interface file
-    let handleExists e | isDoesNotExistError e = return ()
+  -- The Agda.Primitive* and Agda.Builtin* modules.
+  let builtins = filter ((== ".agda") . takeExtension) (dataFiles pd)
+
+  -- Remove all existing .agdai files.
+  forM_ builtins $ \fp -> do
+    let fullpathi = toIFile (ddir </> fp)
+
+        handleExists e | isDoesNotExistError e = return ()
                        | otherwise             = throwIO e
 
     removeFile fullpathi `catch` handleExists
 
-    putStrLn $ "... " ++ fullpath
-    ok <- rawSystem' ddir agda [ "--no-libraries", "--local-interfaces"
-                               , "--ignore-all-interfaces"
-                               , "-Werror"
-                               , fullpath, "-v0"
-                               ]
-    case ok of
-      ExitSuccess   -> return ()
-      ExitFailure _ -> die $ "Error: Failed to typecheck " ++ fullpath ++ "!"
+  -- Type-check all builtin modules (in a single Agda session to take
+  -- advantage of caching).
+  let loadBuiltinCmds = concat
+        [ [ cmd ("Cmd_load " ++ f ++ " []")
+          , cmd "Cmd_no_metas"
+            -- Fail if any meta-variable is unsolved.
+          ]
+        | b <- builtins
+        , let f     = show (ddir </> b)
+              cmd c = "IOTCM " ++ f ++ " None Indirect (" ++ c ++ ")"
+        ]
+  env <- getEnvironment
+  bracket
+    (createProcess
+      ((proc agda
+          [ "--interaction"
+          , "--interaction-exit-on-error"
+          , "--no-libraries"
+          , "--local-interfaces"
+          , "-Werror"
+          , "-v0"
+          ])
+        { std_in        = CreatePipe
+        , std_out       = CreatePipe
+        , delegate_ctlc = True
+                          -- Make Agda look for data files in a
+                          -- certain place.
+        , env           = Just (("Agda_datadir", ddir) : env)
+        }))
+    (\(Just wr, Just rd, _, p) -> do
+       -- Try to let Agda shut down gracefully.
+       hClose wr
+       ok <- waitForProcess p
+       hClose rd
+       case ok of
+         ExitSuccess   -> return ()
+         ExitFailure _ ->
+           die "Error: Failed to typecheck a builtin module!")
+    (\(Just wr, Just _, Nothing, p) -> do
```

Comment:
You never consume the output generated by Agda, since you even throw away the output handle created by `CreatePipe`.  This means that the pipe will block once it is full, and the buffer size is OS-specific.  As a consequence `Setup` might freeze, such as we experience under Windows.

Reply:
I read that you have access to a Windows machine with a suitable build environment. Can you fix this?

Reply:
The following code seems to work under Linux:
```haskell
import Control.Concurrent (forkIO)
```
```haskell
    (\(Just wr, Just rd, Nothing, p) -> do
       mapM_ (flip hSetEncoding utf8) [wr, rd]
       hSetBuffering wr LineBuffering
       -- TODO: Document why the following line is here.
       _ <- forkIO (hGetContents rd >> return ())
       mapM_ (hPutStrLn wr) loadBuiltinCmds)
```